x/participationrewards/keeper/callbacks.go

```diff
@@ -407,3 +414,123 @@ func SetEpochBlockCallback(ctx sdk.Context, k *Keeper, args []byte, query icqtyp
 	k.SetProtocolData(ctx, connectionData.GenerateKey(), &data)
 	return nil
 }
+
+func CrescentPoolUpdateCallback(ctx sdk.Context, k *Keeper, response []byte, query icqtypes.Query) error {
+	var pd liquiditytypes.Pool
+	if err := k.cdc.Unmarshal(response, &pd); err != nil {
+		return err
+	}
+
+	if query.Request[0] != liquiditytypes.PoolKeyPrefix[0] {
+		return errors.New("crescent pool query request has unexpected prefix")
+	}
+
+	poolID := sdk.BigEndianToUint64(query.Request[1:])
+
+	data, ok := k.GetProtocolData(ctx, types.ProtocolDataTypeCrescentPool, fmt.Sprintf("%d", poolID))
+	if !ok {
+		return fmt.Errorf("unable to find protocol data for crescent-types pool/%d", poolID)
+	}
+	ipool, err := types.UnmarshalProtocolData(types.ProtocolDataTypeCrescentPool, data.Data)
+	if err != nil {
+		return err
+	}
+	pool, ok := ipool.(*types.CrescentPoolProtocolData)
+	if !ok {
+		return fmt.Errorf("unable to unmarshal protocol data for crescent-types pool/%d", poolID)
+	}
+	pool.PoolData, err = json.Marshal(pd)
+	if err != nil {
+		return err
+	}
+	pool.LastUpdated = ctx.BlockTime()
+	data.Data, err = json.Marshal(pool)
+	if err != nil {
+		return err
+	}
+	k.SetProtocolData(ctx, pool.GenerateKey(), &data)
+
+	return nil
+}
+
+func CrescentPoolCoinSupplyUpdateCallback(ctx sdk.Context, k *Keeper, response []byte, query icqtypes.Query) error {
+	supplyAmount := sdk.ZeroInt()
+	if err := supplyAmount.Unmarshal(response); err != nil {
+		return err
+	}
+
```

Comment:
```suggestion
if len(query.Request) < 2 {
		k.Logger(ctx).Error("unable to unmarshal balance request, request length is too short")
		return errors.New("account balance icq request must always have a length of at least 2 bytes")
	}
```

should be safe as quicksilver creates the query.Request, but be defensive.

Reply:
added a slight variation